controllers/helper.go

```diff
@@ -40,6 +45,46 @@ const (
 
 var namespace = os.Getenv("NAMESPACE")
 
+type DrainAnnotationPredicate struct {
+	predicate.Funcs
+}
+
+func (DrainAnnotationPredicate) Create(e event.CreateEvent) bool {
+	logger := log.FromContext(context.TODO())
+	if e.Object == nil {
+		logger.Info("Create event: node has no drain annotation", "node", e.Object.GetName())
+		return false
+	}
+
+	if _, hasAnno := e.Object.GetAnnotations()[constants.NodeDrainAnnotation]; hasAnno {
+		logger.Info("Create event: node has no drain annotation", "node", e.Object.GetName())
+		return true
+	}
+	return false
+}
+
+func (DrainAnnotationPredicate) Update(e event.UpdateEvent) bool {
+	logger := log.FromContext(context.TODO())
+	if e.ObjectOld == nil {
+		logger.Info("Update event has no old object to update", "node", e.ObjectOld.GetName())
+		return false
+	}
+	if e.ObjectNew == nil {
+		logger.Info("Update event has no new object for update", "node", e.ObjectNew.GetName())
+		return false
+	}
+
+	oldAnno, hasOldAnno := e.ObjectOld.GetAnnotations()[constants.NodeDrainAnnotation]
+	newAnno, hasNewAnno := e.ObjectNew.GetAnnotations()[constants.NodeDrainAnnotation]
+
+	if !hasOldAnno || !hasNewAnno {
+		logger.Info("Update event: can not compare annotations", "node", e.ObjectNew.GetName())
+		return false
+	}
+
+	return oldAnno == newAnno
```

Comment:
From the doc:
> // Update returns true if the Update event should be processed

So I'd change it to `oldAnno != newAnno`, as the event should be processed if a value change has happened. Right?

Reply:
done